cirq-core/cirq/sim/clifford/act_on_stabilizer_args.py

```diff
@@ -75,19 +75,15 @@ def _act_on_fallback_(
         qubits: Sequence['cirq.Qid'],
         allow_decompose: bool = True,
     ) -> Union[bool, NotImplementedType]:
-        strats = [
-            self._strat_apply_gate,
-            self._strat_apply_mixture,
-        ]
+        if self._strat_apply_gate(action, qubits):
+            return True
         if allow_decompose:
-            strats.append(self._strat_decompose)
-            strats.append(self._strat_act_from_single_qubit_decompose)
-        for strat in strats:
-            result = strat(action, qubits)  # type: ignore
-            if result is True:
+            if self._strat_act_from_single_qubit_decompose(action, qubits):
+                return True
+            if self._strat_apply_mixture(action, qubits):
+                return True
+            if self._strat_decompose(action, qubits):
```

Comment:
Order and nesting got mixed up here - this should be:
```
if self._strat_apply_gate: {...}
if self._strat_apply_mixture: {...}
if allow_decompose:
    if self._strat_decompose: {...}
    if self._strat_act_from_single_qubit_decompose: {...}
```
Ideally, we would also have a test that catches this (i.e. fails with this ordering, but succeeds with the correct order)

Reply:
This was roughly intentional. see the 2nd comment on the thread.

Reply:
To be precise, the following is what happens if we preserve the exact same calls:

```python
if self._strat_apply_gate():
  return True
if self._strat_apply_mixture():
  return True
if allow_decompose:
  if self._strat_decompose():
    return True
  if self._strat_act_from_single_qubit_decompose():
    return True  # <- this is never hit because `_strat_apply_mixture()` handles a strict superset of what this can handle.
```

So, we can delete the last one, but it implies `_strat_apply_mixture` belongs in the `allow_decompose` block, since it calls `_strat_act_from_single_qubit_decompose` under the hood.

Reply:
Also xref #1657. If we could have mixtures be returned as operations, the `_strat_apply_mixture` would *not* be a decomposing strategy. (Or, we could have a separate decomposing and non-decomposing option for it). This would be particularly useful for depolarizing channels, which are a mixture of Paulis, that the clifford simulators can handle directly.

Reply:
Also tangentially, there's probably an opportunity to record a channel_measurement in `_strat_apply_mixture`.

Reply:
There's a bunch of odd corners here, but first: the intent of `allow_decompose` is to control whether this method should decompose _user-defined gates_. I'll use "decomposing strategy" below to refer to strategies which require `allow_decompose=True`.

1. When called on a mixture (not a gate), `_strat_apply_mixture` generates a `MatrixGate` to allow reuse of `_strat_act_from_single_qubit_decompose`. Since this gate was not user-generated (it's just a wrapper around a unitary), this is **not** a decomposing strategy.
2. When called on a _gate_, `_strat_apply_mixture` first converts the gate to a mixture with something like `mixture([(1.0, op.unitary)])`, then generates a `MatrixGate` from that. This **is** a decomposing strategy, since the result is identical to calling `_strat_act_from_single_qubit_decompose` on the original gate.
3. _However_, case (2) is only encountered after calling `_strat_apply_gate` on the gate; if that method succeeds, then case (2) does not trigger.

What this boils down to is: `_strat_apply_mixture` can be called outside the `allow_decompose` check if the input is _actually_ a mixture. For gates it would need to be called inside the check, but since `_strat_act_from_single_qubit_decompose` has the same behavior we can just skip it instead.

This produces:
```
if self._strat_apply_gate(...): {...}
if (action is a mixture, not a gate):
    if self._strat_apply_mixture(...): {...}
if allow_decompose:
    if self._strat_decompose: {...}
    if self._strat_act_from_single_qubit_decompose: {...}
```
Since gates qualify as mixtures, the `action is a mixture, not a gate` check must be made with some caution.

Reply:
Okay I see. Actually I am going to abandon at this point because we may consider going the other direction and putting the core fallback logic in the base class, so that have the subclasses only have to return the strategy array. That would require undoing everything we did here, so I feel like "wait and see" is the best approach for now.